src/compiler/utils/string_with_map.ts

```diff
@@ -0,0 +1,189 @@
+import { encode } from "sourcemap-codec";
+
+type MappingSegment = [ number ] | [ number, number, number, number ] |  [ number, number, number, number, number ]
+
+type SourceMappings = {
+    sources: string[];
+    names: string[];
+    mappings: MappingSegment[][];
+}
+
+type SourceLocation = {
+    line: number;
+    column: number;
+}
+
+function get_end_location(s: string): SourceLocation {
+    const parts = s.split('\n');
+    return {
+        line: parts.length - 1,
+        column: parts[parts.length - 1].length - 1
+    };
+}
+
+
+export function offset_source_location(offset: SourceLocation, map: SourceMappings): SourceMappings {
+ 
+    const new_mappings = map.mappings.map(line => line.map(seg => {
+        if (seg.length < 3) return seg;
+        const new_seg = seg.slice() as MappingSegment;
+        new_seg[2] = new_seg[2] + offset.line;
+        return new_seg;
+    }));
+
+    // first line has column altered
+    if (new_mappings.length > 0) {
+        new_mappings[0] = new_mappings[0].map(seg => {
+            if (seg.length < 4) return seg;
+            const newSeg = seg.slice() as MappingSegment;
+            newSeg[3] = newSeg[3] + offset.column;
+            return newSeg;
+        });
+    }
+    
+    return {
+        sources: map.sources,
+        mappings: new_mappings
+    } as SourceMappings;
+}
+
+
+
+function merge_tables<T>( original: T[], extended: T[]): { table: T[]; new_idx: number[] } {
+    const table = original.slice();
+        const new_idx = [];
+        for (let j = 0; j < original.length; j++) {
+            const current = extended[j];
+            const existing = table.indexOf(current);
+            if (existing < 0) {
+                table.push(current);
+                new_idx[j] = table.length - 1;
+            } else {
+                new_idx[j] = existing;
+            }
+        }
+    return { table, new_idx }; 
+}
+
+
+
+export class GeneratedStringWithMap {
+    readonly generated: string;
+    readonly map: SourceMappings;
+
+    constructor(generated: string , map: SourceMappings) {
+        this.generated = generated;
+        this.map = map;
+    }
+ 
+    as_sourcemap() {
+        return {
+            version: 3,
+            sources: this.map.sources,
+            names: [],
+            mappings: encode(this.map.mappings as any)
+        };
+    }
+    
+    concat(other: GeneratedStringWithMap): GeneratedStringWithMap {
+        // if one is empty, return the other
+        if (this.generated.length == 0) return other;
+        if (other.generated.length == 0) return this;
+    
+        //combine sources
```

Comment:
there's a mix here with there sometimes being a space after `//` and sometimes not. I'd probably always include a space for consistency